daemon/api_snaps_test.go

```diff
@@ -1078,3 +1078,886 @@ func (s *snapsSuite) TestMapLocalOfTryResolvesSymlink(c *check.C) {
+	inst := &daemon.SnapInstruction{
+		Action: "install",
+		Snaps:  []string{""},
+	}
+
+	st := d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+	_, _, err := inst.Dispatch()(inst, st)
+	c.Check(err, check.ErrorMatches, "cannot install snap with empty name")
+}
+
+func (s *snapsSuite) TestInstallOnNonDevModeDistro(c *check.C) {
+	s.testInstall(c, false, snapstate.Flags{}, snap.R(0))
+}
+func (s *snapsSuite) TestInstallOnDevModeDistro(c *check.C) {
+	s.testInstall(c, true, snapstate.Flags{}, snap.R(0))
+}
+func (s *snapsSuite) TestInstallRevision(c *check.C) {
+	s.testInstall(c, false, snapstate.Flags{}, snap.R(42))
+}
+
+func (s *snapsSuite) testInstall(c *check.C, forcedDevmode bool, flags snapstate.Flags, revision snap.Revision) {
+	calledFlags := snapstate.Flags{}
+	installQueue := []string{}
+	restore := sandbox.MockForceDevMode(forcedDevmode)
+	defer restore()
+
+	defer daemon.MockSnapstateInstall(func(ctx context.Context, s *state.State, name string, opts *snapstate.RevisionOptions, userID int, flags snapstate.Flags) (*state.TaskSet, error) {
+		calledFlags = flags
+		installQueue = append(installQueue, name)
+		c.Check(revision, check.Equals, opts.Revision)
+
+		t := s.NewTask("fake-install-snap", "Doing a fake install")
+		return state.NewTaskSet(t), nil
+	})()
+
+	d := s.daemonWithFakeSnapManager(c)
+
+	var buf bytes.Buffer
+	if revision.Unset() {
+		buf.WriteString(`{"action": "install"}`)
+	} else {
+		fmt.Fprintf(&buf, `{"action": "install", "revision": %s}`, revision.String())
+	}
+	req, err := http.NewRequest("POST", "/v2/snaps/some-snap", &buf)
+	c.Assert(err, check.IsNil)
+
+	rsp := s.req(c, req, nil).(*daemon.Resp)
+
+	c.Assert(rsp.Type, check.Equals, daemon.ResponseTypeAsync)
+
+	st := d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+	chg := st.Change(rsp.Change)
+	c.Assert(chg, check.NotNil)
+
+	c.Check(chg.Tasks(), check.HasLen, 1)
+
+	st.Unlock()
+	s.waitTrivialChange(c, chg)
+	st.Lock()
+
+	c.Check(chg.Status(), check.Equals, state.DoneStatus)
+	c.Check(calledFlags, check.Equals, flags)
+	c.Check(err, check.IsNil)
+	c.Check(installQueue, check.DeepEquals, []string{"some-snap"})
+	c.Check(chg.Kind(), check.Equals, "install-snap")
+	c.Check(chg.Summary(), check.Equals, `Install "some-snap" snap`)
+}
+
+func (s *snapsSuite) TestInstallUserAgentContextCreated(c *check.C) {
+	defer daemon.MockSnapstateInstall(func(ctx context.Context, st *state.State, name string, opts *snapstate.RevisionOptions, userID int, flags snapstate.Flags) (*state.TaskSet, error) {
+		s.ctx = ctx
+		t := st.NewTask("fake-install-snap", "Doing a fake install")
+		return state.NewTaskSet(t), nil
+	})()
+
+	s.daemonWithFakeSnapManager(c)
+
+	var buf bytes.Buffer
+	buf.WriteString(`{"action": "install"}`)
+	req, err := http.NewRequest("POST", "/v2/snaps/some-snap", &buf)
+	req.RemoteAddr = "pid=100;uid=0;socket=;"
+	c.Assert(err, check.IsNil)
+	req.Header.Add("User-Agent", "some-agent/1.0")
+
+	s.vars = map[string]string{"name": "some-snap"}
```

Comment:
(nipick) Is this needed? It seems the request does that for the relevant pieces, no?